app/controllers/admin/badges_controller.rb

```diff
@@ -33,6 +35,38 @@ def new
   def show
   end
 
+  def award
+  end
+
+  def mass_award
+    csv_file = params.permit(:file).fetch(:file, nil)
+    badge = Badge.find_by(id: params[:badge_id])
+    raise Discourse::InvalidParameters if csv_file.try(:tempfile).nil? || badge.nil?
+
+    batch_number = 1
+    batch = []
+
+    File.open(csv_file) do |csv|
+      csv.each_line do |email_line|
+        batch.concat CSV.parse_line(email_line)
+
+        # Split the emails in batches of 200 elements.
+        full_batch = csv.lineno % (BadgeGranter::MAX_ITEMS_FOR_DELTA * batch_number) == 0
```

Comment:
I wonder if 200 emails will be too large as a redis value for sidekiq? Might be ok, but we should definitely test with batches larger than 200 and perhaps tune it.